Cancelling MVS/TSO command causes error

**Describe the bug**

When pressing escape in the text input field for entering either an MVS or TSO command, there are errors that show up. This is not expected as we are cancelling the "issue command" command, not submitting the empty or un-finished command.

Cancelling issue MVS command error:
```
[2025/01/29 11:07:49] [TRACE] ZoweCommandProvider.issueCommand called.
[2025/01/29 11:07:49] [ERROR] Error: Rest API failure with HTTP(S) status 400
{
  errorDetails: s [Error]: Rest API failure with HTTP(S) status 400
      at m.populateError (c:\Users\JAWATERS\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0\out\src\vendors.extension.js:416:284)
      at x (c:\Users\JAWATERS\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0\out\src\vendors.extension.js:403:6976)
      at m.onEnd (c:\Users\JAWATERS\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0\out\src\vendors.extension.js:403:7217)
      at IncomingMessage.<anonymous> (c:\Users\JAWATERS\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0\out\src\vendors.extension.js:403:5394)
      at IncomingMessage.emit (node:events:530:35)
      at endReadableNT (node:internal/streams/readable:1698:12)
      at processTicksAndRejections (node:internal/process/task_queues:82:21) {
    mDetails: {
      msg: 'Rest API failure with HTTP(S) status 400',
      causeErrors: `{"reason":"Can not find 'cmd' field in request body OR value of 'cmd' is empty. No command will be issued!","return-code":1,"reason-code":13}`,
      source: 'http',
      errorCode: 400,
      protocol: 'https',
      port: port,
      host: 'host',
      basePath: '',
      httpStatus: 400,
      errno: undefined,
      syscall: undefined,
      payload: { cmd: '' },
      headers: [
        { 'Content-Type': 'application/json' },
        { 'X-CSRF-ZOSMF-HEADER': true }
      ],
      resource: '/zosmf/restconsoles/consoles/defcn',
      request: 'PUT',
      additionalDetails: 'Received HTTP(S) error 400 = Bad Request.\n' +
        '\n' +
        'Protocol:          https\n' +
        'Host:             host\n' +
        'Port:              port\n' +
        'Base Path:         \n' +
        'Resource:          /zosmf/restconsoles/consoles/defcn\n' +
        'Request:           PUT\n' +
        'Headers:           [{"Content-Type":"application/json"},{"X-CSRF-ZOSMF-HEADER":true}]\n' +
        "Payload:           { cmd: '' }\n" +
        'Auth type:         basic\n' +
        'Allow Unauth Cert: true'
    },
    mMessage: 'Rest API failure with HTTP(S) status 400'
  },
  apiType: 'cmd',
  profile: undefined
}
```

Cancelling issue TSO command error:
```
[2025/01/29 11:09:35] [TRACE] ZoweCommandProvider.issueCommand called.
[2025/01/29 11:09:35] [ERROR] Error: Expect Error: Imperative API Error No command text was provided.
{
  errorDetails: o [Error]: Expect Error: Imperative API Error No command text was provided.
      at Function.toNotBeEqual (c:\Users\JAWATERS\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0\out\src\vendors.extension.js:316:5334)
      at Function.validateNotEmptyString (c:\Users\JAWATERS\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0\out\src\vendors.extension.js:584:7372)
      at Function.<anonymous> (c:\Users\JAWATERS\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0\out\src\vendors.extension.js:583:293)
      at Generator.next (<anonymous>)
      at b (c:\Users\JAWATERS\.vscode\extensions\zowe.vscode-extension-for-zowe-3.1.0\out\src\vendors.extension.js:582:6812)
      at processTicksAndRejections (node:internal/process/task_queues:95:5) {
    mDetails: { msg: 'Imperative API Error No command text was provided.' },
    mMessage: 'Expect Error: Imperative API Error No command text was provided.'
  },
  apiType: 'cmd',
  profile: undefined
}
```

**To Reproduce**

Steps to reproduce the behavior:
1. Log into zosmf profile
2. Right click tree and click "issue MVS/TSO command"
3. Press escape while in the "enter command" text field
4. See error

**Expected behavior**

No error, or perhaps a message box saying command cancelled or similar.

**Desktop (please complete the following information):**

- OS: Windows
- Zowe Explorer Version: v3.1.0
- (Optional) Zowe CLI Version:
- (Optional) Are you using Secure Credential Store? Yes

--- a/packages/zowe-explorer/CHANGELOG.md
+++ b/packages/zowe-explorer/CHANGELOG.md
@@ -15,6 +15,7 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 
 - Fixed an issue where a TypeError occurred when applying VS Code proxy settings to an invalid session. [#3425](https://github.com/zowe/zowe-explorer-vscode/issues/3425)
 - Fixed issue where the 'Delete' key binding for the USS tree returns a 'contextValue' error. [#2796](https://github.com/zowe/zowe-explorer-vscode/issues/2796)
+- Fixed issue where cancelling Unix, MVS or TSO command still submits the command. [#3422](https://github.com/zowe/zowe-explorer-vscode/issues/3422)
 
 ## `3.1.0`
 

--- a/packages/zowe-explorer/__tests__/__unit__/commands/MvsCommandHandler.unit.test.ts
+++ b/packages/zowe-explorer/__tests__/__unit__/commands/MvsCommandHandler.unit.test.ts
@@ -332,7 +332,7 @@ describe("mvsCommandActions unit testing", () => {
             ignoreFocusOut: true,
             placeHolder: "Select the profile to use to submit the MVS command",
         });
-        expect(showInformationMessage.mock.calls.length).toBe(0);
+        expect(showInformationMessage.mock.calls.length).toBe(1);
     });
 
     it("tests the issueMvsCommand function user escapes the MVS command box", async () => {
@@ -371,7 +371,7 @@ describe("mvsCommandActions unit testing", () => {
             placeHolder: "Select the profile to use to submit the MVS command",
         });
         expect(showInputBox.mock.calls.length).toBe(1);
-        expect(showInformationMessage.mock.calls.length).toBe(0);
+        expect(showInformationMessage.mock.calls.length).toBe(1);
     });
 
     it("tests the issueMvsCommand function user starts typing a value in quick pick", async () => {

--- a/packages/zowe-explorer/__tests__/__unit__/commands/TsoCommandHandler.unit.test.ts
+++ b/packages/zowe-explorer/__tests__/__unit__/commands/TsoCommandHandler.unit.test.ts
@@ -371,7 +371,7 @@ describe("TsoCommandHandler unit testing", () => {
             placeHolder: "Select the profile to use to submit the TSO command",
         });
         expect(showInputBox.mock.calls.length).toBe(1);
-        expect(showInformationMessage.mock.calls.length).toBe(0);
+        expect(showInformationMessage.mock.calls.length).toBe(1);
     });
 
     it("tests the issueTsoCommand function user starts typing a value in quick pick", async () => {

--- a/packages/zowe-explorer/__tests__/__unit__/commands/UnixCommandHandler.unit.test.ts
+++ b/packages/zowe-explorer/__tests__/__unit__/commands/UnixCommandHandler.unit.test.ts
@@ -359,7 +359,7 @@ describe("UnixCommand Actions Unit Testing", () => {
             placeHolder: "Select the profile to use to submit the Unix command",
         });
         expect(showInputBox.mock.calls.length).toBe(1);
-        expect(showInformationMessage.mock.calls.length).toBe(0);
+        expect(showInformationMessage.mock.calls.length).toBe(1);
     });
 
     it("tests the issueUnixCommand function user escapes the commandbox", async () => {
@@ -389,7 +389,7 @@ describe("UnixCommand Actions Unit Testing", () => {
             placeHolder: "Select the profile to use to submit the Unix command",
         });
         expect(showInputBox.mock.calls.length).toBe(2);
-        expect(showInformationMessage.mock.calls.length).toBe(0);
+        expect(showInformationMessage.mock.calls.length).toBe(1);
     });
 
     it("tests the issueUnixCommand function - issueUnixCommand throws an error", async () => {

--- a/packages/zowe-explorer/__tests__/__unit__/commands/ZoweCommandProvider.unit.test.ts
+++ b/packages/zowe-explorer/__tests__/__unit__/commands/ZoweCommandProvider.unit.test.ts
@@ -109,6 +109,8 @@ describe("ZoweCommandProvider Unit Tests", () => {
                 runCommand: jest.fn().mockRejectedValue(testError),
             };
             const testProfile: any = { name: "test", profile: { user: "firstName", password: "12345" } };
+            const showInformationMessage = jest.fn();
+            Object.defineProperty(vscode.window, "showInformationMessage", { value: showInformationMessage });
 
             it("should not create a terminal if the profile or the command is undefined", async () => {
                 const createTerminal = jest.fn().mockReturnValue({ show: jest.fn() });
@@ -118,6 +120,14 @@ describe("ZoweCommandProvider Unit Tests", () => {
                 await ZoweCommandProvider.prototype.issueCommand.call(mockCmdProvider, testProfile, null);
                 await ZoweCommandProvider.prototype.issueCommand.call(mockCmdProvider, testProfile, undefined);
                 expect(createTerminal).not.toHaveBeenCalled();
+            });
+
+            it("should not create a terminal if user escapes the input box", async () => {
+                const createTerminal = jest.fn().mockReturnValue({ show: jest.fn() });
+                Object.defineProperty(vscode.window, "createTerminal", { value: createTerminal, configurable: true });
+                await ZoweCommandProvider.prototype.issueCommand.call(mockCmdProvider, testProfile, "");
+                expect(createTerminal).not.toHaveBeenCalled();
+                expect(showInformationMessage.mock.calls.length).toBe(1);
             });
 
             it("should create an integrated terminal", async () => {

--- a/packages/zowe-explorer/__tests__/__unit__/commands/ZoweCommandProvider.unit.test.ts
+++ b/packages/zowe-explorer/__tests__/__unit__/commands/ZoweCommandProvider.unit.test.ts
@@ -122,7 +122,10 @@ describe("ZoweCommandProvider Unit Tests", () => {
                 expect(createTerminal).not.toHaveBeenCalled();
             });
 
-            it("should not create a terminal if user escapes the input box", async () => {
+            it("should not create a terminal if user escapes the input box and useIntegratedTerminals is false", async () => {
+                jest.spyOn(SettingsConfig, "getDirectValue").mockImplementation(
+                    (setting) => setting !== Constants.SETTINGS_COMMANDS_INTEGRATED_TERMINALS
+                );
                 const createTerminal = jest.fn().mockReturnValue({ show: jest.fn() });
                 Object.defineProperty(vscode.window, "createTerminal", { value: createTerminal, configurable: true });
                 await ZoweCommandProvider.prototype.issueCommand.call(mockCmdProvider, testProfile, "");

--- a/packages/zowe-explorer/src/commands/ZoweCommandProvider.ts
+++ b/packages/zowe-explorer/src/commands/ZoweCommandProvider.ts
@@ -61,6 +61,10 @@ export abstract class ZoweCommandProvider {
         if (profile == null || command == null) {
             return;
         }
+        if (command.length === 0) {
+            Gui.showMessage(this.operationCancelled);
+            return;
+        }
         try {
             const iTerms = SettingsConfig.getDirectValue(Constants.SETTINGS_COMMANDS_INTEGRATED_TERMINALS);
             if (iTerms) {

--- a/packages/zowe-explorer/src/commands/ZoweCommandProvider.ts
+++ b/packages/zowe-explorer/src/commands/ZoweCommandProvider.ts
@@ -61,10 +61,6 @@ export abstract class ZoweCommandProvider {
         if (profile == null || command == null) {
             return;
         }
-        if (command.length === 0) {
-            Gui.showMessage(this.operationCancelled);
-            return;
-        }
         try {
             const iTerms = SettingsConfig.getDirectValue(Constants.SETTINGS_COMMANDS_INTEGRATED_TERMINALS);
             if (iTerms) {
@@ -129,6 +125,10 @@ export abstract class ZoweCommandProvider {
                 this.terminal = vscode.window.createTerminal({ name: `(${profile.name}) ${this.terminalName}`, pty: this.pseudoTerminal });
                 this.terminal.show();
             } else {
+                if (command.length === 0) {
+                    Gui.showMessage(this.operationCancelled);
+                    return;
+                }
                 this.outputChannel ??= Gui.createOutputChannel(this.terminalName);
                 this.outputChannel.appendLine(this.formatCommandLine(command, profile));
                 const response = await Gui.withProgress(

